pennylane/workflow/qnode.py

```diff
@@ -110,15 +110,33 @@ def _to_qfunc_output_type(
     if has_partitioned_shots:
         return tuple(_to_qfunc_output_type(r, qfunc_output, False) for r in results)
 
+    qfunc_output_leaves, qfunc_output_structure = qml.pytrees.flatten(
+        qfunc_output, is_leaf=lambda obj: isinstance(obj, (qml.measurements.MeasurementProcess))
+    )
+    num_of_measurements = len(qfunc_output_leaves)
+
+    results_leaves, results_structure = qml.pytrees.flatten(results)
+    num_of_results = len(results_leaves)
+
     # Special case of single Measurement in a list
     if isinstance(qfunc_output, list) and len(qfunc_output) == 1:
         results = [results]
 
-    # If the return type is not tuple (list or ndarray) (Autograd and TF backprop removed)
-    if isinstance(qfunc_output, (tuple, qml.measurements.MeasurementProcess)):
+    if isinstance(qfunc_output, qml.measurements.MeasurementProcess):
         return results
 
-    return type(qfunc_output)(results)
+    if num_of_measurements != num_of_results or (

```

Comment:
Could you add a brief inline comment about what is going on here? In the context of this PR it's easy to see, but I think without the context, it might be harder to interpret if we are reading through this fine a few months from now.